compiler/internal/dce/info.go

```diff
@@ -0,0 +1,108 @@
+package dce
+
+import (
+	"fmt"
+	"go/types"
+	"sort"
+	"strings"
+
+	"github.com/gopherjs/gopherjs/compiler/typesutil"
+)
+
+// Info contains information used by the dead-code elimination (DCE) logic to
+// determine whether a declaration is alive or dead.
+type Info struct {
+
+	// alive indicates if the declaration is marked as alive
+	// and will not be eliminated.
+	alive bool
+
+	// importPath is the package path of the package the declaration is in.
+	importPath string
+
+	// Symbol's identifier used by the dead-code elimination logic, not including
+	// package path. If empty, the symbol is assumed to be alive and will not be
+	// eliminated. For methods it is the same as its receiver type identifier.
+	objectFilter string
+
+	// The second part of the identified used by dead-code elimination for methods.
+	// Empty for other types of symbols.
+	methodFilter string
```

Comment:
optional: I am happy to merge this RP as-is to limit the amount of changes in the logic, but it may be a worthwhile refactoring to consider.

It may be worthwhile grouping these three strings into a `dce.Key` type, and moving logic such as key derivation from a `types.Object` and key comparison into that type. I think that way we could use the same key representation everywhere in the code. Currently, for example, `dceInfo.objectFilter` is subtly different from `Info.objectFilter` because the former is qualified by the import path. I have a hunch that having `dce.Key` type may make things a little clearer instead of having to rely on untyped strings.

Anyway, this is just food for thought, I am okay with merging as is, if you prefer.

Reply:
So, I did this abstraction prior to really understanding the DCE very well. I was trying to make it just a simple abstraction with minimal changes. However, I think if I update towards what I have in the change to handle instances PR, #1340, it will be closer to what you were thinking, but still without a key object.

For example, I later realized it didn't make sense to keep `importPath` separate, so the names are back to just `objectFilter` and `methodFilter`. I also added a `filters.go` that does all the name generation stuff. Eventually the dependencies and names both use names returned from the same filter generation function. The filter generation does get somewhat complicated so it might be better to rethink creating a `Key` later when reviewing #1340.